applications/StructuralMechanicsApplication/custom_advanced_constitutive/plane_stress_d_plus_d_minus_damage_masonry_2d.cpp

```diff
@@ -7,752 +7,1076 @@
-	const double principal_stress_1 = rPrincipalStressVector[0];
-	const double principal_stress_2 = rPrincipalStressVector[1];
-	const double smax_macaulay = std::max(principal_stress_1, 0.0);
-	if (principal_stress_2 < 0.0){
-		rEquivalentStress = alpha_factor * (alpha*I1 + std::sqrt(3.0 * J2) + beta * shear_compression_reductor * smax_macaulay);
+	if (std::abs(eigen_values_matrix(0, 0) - eigen_values_matrix(1, 1)) > 0.0) {
+		factor_12 *= 2.0;
+		factor_12 /= eigen_values_matrix(0, 0) - eigen_values_matrix(1, 1);
+	} else {
+		factor_12 = 1.0;
 	}
+	noalias(projection_tensor_tension) += factor_12 * outer_prod(projection_vector_cross, projection_vector_cross);
+#endif //PROJECTION_OPERATOR_CERVERA_2003
+
+#ifdef PROJECTION_OPERATOR_CERVERA_2017
+/*
+Theory from: 	"An Energy-Equivalent d+/d- Damage Model with
+				Enhanced Microcrack Closure-Reopening
+				Capabilities for Cohesive-Frictional Materials"
+				Miguel Cervera & Claudia Tesei
+				2017 (page 7/30)
+*/
+	array_1d<double,3> projection_vector_12;
+	Matrix projection_tensor_12;
+	array_1d<double,3> projection_vector_21;
+	Matrix projection_tensor_21;
+	projection_tensor_12 = outer_prod(eigen_vector_1, eigen_vector_2);
+	projection_tensor_21 = outer_prod(eigen_vector_2, eigen_vector_1);
+
+	array_1d<double,3> projection_vector_cross;
+	Matrix projection_tensor_cross;
+	projection_tensor_cross = 0.5 * (projection_tensor_12 + projection_tensor_21);
+	projection_vector_cross = MathUtils<double>::StressTensorToVector(projection_tensor_cross);
+
+	noalias(projection_tensor_tension) += (HEAVISIDE(eigen_values_matrix(0, 0)) + HEAVISIDE(eigen_values_matrix(1, 1))) * 
+										  outer_prod(projection_vector_cross, projection_vector_cross);
+#endif //PROJECTION_OPERATOR_CERVERA_2017
+
+	noalias(projection_tensor_compression) = IdentityMatrix(3,3) - projection_tensor_tension;
 }
+
 /***********************************************************************************/
 /***********************************************************************************/
-void DamageDPlusDMinusMasonry2DLaw::IntegrateStressVectorTension(
-	array_1d<double,3>& rPredictiveStressVector,
-    const double UniaxialStress,
-    double& rDamage,
-    double& rThreshold,
-    ConstitutiveLaw::Parameters& rValues,
-    const double CharacteristicLength)
+void DamageDPlusDMinusMasonry2DLaw::CalculateEquivalentStressTension(CalculationData& data, double& UniaxialStressTension)
 {
-	double damage_parameter;
-	this->CalculateDamageParameterTension(rValues, damage_parameter, CharacteristicLength);
-	this->CalculateExponentialDamageTension(UniaxialStress, rThreshold, damage_parameter, CharacteristicLength, rValues, rDamage);
+	UniaxialStressTension = 0.0;
+	if(data.PrincipalStressVector(0) > 0.0){
+		if (data.TensionYieldModel == 0) {
+			// Lubliner Yield Criteria 
+			const double yield_compression 		 	= 	data.YieldStressCompression;
+			const double yield_tension 				= 	data.YieldStressTension;

```

Comment:
something weird with the indentations?

Reply:
I dont know why it appears that weird here. For me it looks ok when I open the file in visualstudiocode

Reply:
It is because you are using tabs instead of the 4 spaces. Somehow github deals different with tabs than visual studio and other text editors.